src/big/big_rational.cr

```diff
@@ -139,8 +159,19 @@ struct BigRational < Number
     BigRational.new { |mpq| LibGMP.mpq_abs(mpq, self) }
   end
 
-  def hash
-    to_f64.hash
+  private HASH_MODULUS_RAT = BigRational.new((1_u64 << HASH_BITS) - 1)
+
+  def hash_normalize
+    # more exact version of `remainder(hash_modulus).to_f.hash_normalize`
+    rem = self
+    if self >= HASH_MODULUS_RAT
+      num = numerator
+      denom = denominator
+      div = num.tdiv(denom)
+      floor = div.tdiv(hash_modulus)
+      rem -= floor * hash_modulus
+    end
+    BigFloat.new(rem).hash_normalize
```

Comment:
It looks like BigFloat.new(self) everywhere should be replaced with `to_big_f` with copy of `to_big_f` too. 

Code will be more readable and to strange require needed.